pkg/dataobj/internal/dataset/reader.go

```diff
@@ -125,38 +126,79 @@ func (r *Reader) Read(ctx context.Context, s []Row) (n int, err error) {
 	}
 	r.dl.SetReadRange(readRange)
 
-	count, err := r.inner.ReadColumns(ctx, r.dl.PrimaryColumns(), s[:readSize])
-	if err != nil && !errors.Is(err, io.EOF) {
-		return n, err
-	} else if count == 0 && errors.Is(err, io.EOF) {
-		return 0, io.EOF
-	}
-
 	var primaryColumnBytes int64
 	var primaryColumnPostFilterBytes int64
 	var totalBytesAfterFill int64
+	var rowsRead int
 	var passCount int // passCount tracks how many rows pass the predicate.
 
-	if r.opts.Predicate == nil {
-		// If there's no predicate, all rows are valid.
+	// If there are no predicates, read all columns in the dataset
+	if len(r.opts.Predicates) == 0 {
+		count, err := r.inner.ReadColumns(ctx, r.dl.PrimaryColumns(), s[:readSize])
+		if err != nil && !errors.Is(err, io.EOF) {
+			return n, err
+		} else if count == 0 && errors.Is(err, io.EOF) {
+			return 0, io.EOF
+		}
+
+		rowsRead = count
 		passCount = count
+
 		for i := range count {
 			primaryColumnBytes += s[i].Size()
 		}
 	} else {
-		for i := range count {
-			size := s[i].SizeOfColumns(r.primaryColumnIndexes)
+		// sequentially apply rest of the predicates.
+		for i, p := range r.opts.Predicates {
+			columns, idxs, err := r.predicateColumns(p)
```

Comment:
Calling this for every predicate per read call is fine for now, but let's keep an eye out for whether we need to cache these results depending on how many read calls we end up having 👍